staker/block_validator.go

```diff
@@ -805,19 +808,70 @@ func (v *BlockValidator) advanceValidations(ctx context.Context) (*arbutil.Messa
 	v.reorgMutex.RLock()
 	defer v.reorgMutex.RUnlock()
 
+	for {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		pos := v.validated()
+		if pos >= v.recordSent() {
+			log.Trace("advanceValidations: nothing to validate", "pos", pos)
+			return nil, nil
+		}
+		validationStatus, found := v.validations.Load(pos)
+		if !found {
+			return nil, fmt.Errorf("not found entry for pos %d", pos)
+		}
+		if validationStatus.getStatus() != ValidationDone {
+			log.Trace("advanceValidations: validation not done", "pos", pos, "status", validationStatus.getStatus())
+			return nil, nil
+		}
+		if validationStatus.DoneEntry.Start != v.lastValidGS {
+			log.Warn("Validation entry has wrong start state", "pos", pos, "start", validationStatus.DoneEntry.Start, "expected", v.lastValidGS)
+			validationStatus.Cancel()
+			return &pos, nil
+		}
+		if !validationStatus.DoneEntry.Success {
+			v.possiblyFatal(fmt.Errorf("validation: failed entry pos %d, start %v", pos, validationStatus.DoneEntry.Start))
+			return &pos, nil // if not fatal - retry
+		}
+		err := v.writeLastValidated(validationStatus.DoneEntry.End, validationStatus.DoneEntry.WasmModuleRoots)
+		if err != nil {
+			log.Error("failed writing new validated to database", "pos", pos, "err", err)
+		}
+		go v.recorder.MarkValid(pos, v.lastValidGS.BlockHash)
+		atomicStorePos(&v.validatedA, pos+1, validatorMsgCountValidatedGauge)
+		v.validations.Delete(pos)
+		nonBlockingTrigger(v.createNodesChan)
+		nonBlockingTrigger(v.sendRecordChan)
+		nonBlockingTrigger(v.sendValidationsChan)
+		v.testingProgressMadeMutex.Lock()
+		if v.testingProgressMadeChan != nil {
+			nonBlockingTrigger(v.testingProgressMadeChan)
+		}
+		v.testingProgressMadeMutex.Unlock()
+
+		log.Trace("result validated", "count", v.validated(), "blockHash", v.lastValidGS.BlockHash)
+	}
+}
+
+// return val:
+// *MessageIndex - pointer to bad entry if there is one (requires reorg)
+func (v *BlockValidator) sendValidations(ctx context.Context) (*arbutil.MessageIndex, error) {
+	v.reorgMutex.RLock()
+	defer v.reorgMutex.RUnlock()
+
 	wasmRoots := v.GetModuleRootsToValidate()
-	pos := v.validated() - 1 // to reverse the first +1 in the loop
-validationsLoop:
 	for {
 		if ctx.Err() != nil {
 			return nil, ctx.Err()
 		}
-		v.valLoopPos = pos + 1
 		v.reorgMutex.RUnlock()
 		v.reorgMutex.RLock()
-		pos = v.valLoopPos
-		if pos >= v.recordSent() {
-			log.Trace("advanceValidations: nothing to validate", "pos", pos)
+		pos := v.lastValidationSent()
+		if pos > v.validated()+arbutil.MessageIndex(v.config().ValidationSentLimit)-1 {
+			return nil, nil
+		}
+		if pos > v.recordSent()-1 {
```

Comment:
nit (not important/mandatory): I prefer if `a >= b` over `if a > b - 1`

Reply:
Done.